### PR TITLE
retry webhook updates on failure

### PR DESCRIPTION
**Pull request by ayj**
The self-managed webhook reconciliation does not retry when patch/update fails. This mostly likely happens if the kube-apiserver is temporarily unavailable, e.g. during upgrade, transient network failure. Mitigate this problem by retrying failed patch/updates every second until success.

(not tested yet)

### CODE REVIEW

**Comment by ayj** [comment]
> Could not setup KinD environment. Something wrong with KinD setup. Exporting logs.

/test integ-pilot-k8s-tests_istio

File: sidecar-injector/cmd/sidecar-injector/main.go [hunk 1]
```diff
@@ -164,12 +166,16 @@ func patchCertLoop(stopCh <-chan struct{}) error {
 		0,
 		cache.ResourceEventHandlerFuncs{
 			UpdateFunc: func(oldObj, newObj interface{}) {
-				config := newObj.(*v1beta1.MutatingWebhookConfiguration)
-				for i, w := range config.Webhooks {
-					if w.Name == flags.webhookName && !bytes.Equal(config.Webhooks[i].ClientConfig.CABundle, caCertPem) {
-						log.Infof("Detected a change in CABundle, patching MutatingWebhookConfiguration again")
-						shouldPatch <- struct{}{}
-						break
+				oldConfig := newObj.(*v1beta1.MutatingWebhookConfiguration)
```
Comment (howardjohn) [inline sidecar-injector/cmd/sidecar-injector/main.go]: this is the same as newConfig?

Comment (ayj) [reply sidecar-injector/cmd/sidecar-injector/main.go]: copy/paste error. Fixed now - thanks!

File: galley/pkg/crd/validation/config.go [hunk 2]
```diff
@@ -390,14 +403,32 @@ func (whc *WebhookConfigController) reconcile(stopCh <-chan struct{}) {
 			// rebuild the desired configuration and reconcile with the
 			// existing configuration.
 			if err := whc.rebuildWebhookConfig(); err == nil {
-				whc.createOrUpdateWebhookConfig()
+				if retry := whc.createOrUpdateWebhookConfig(); retry {
+					configTimerC = time.After(retryUpdateAfterFailureTimeout)
+					if !retrying {
+						retrying = true
+						log.Info("webhook create/update failed - retrying")
```
Comment (sushicw) [inline galley/pkg/crd/validation/config.go]: Should we print this every retry, or alternately include the retry interval here?

Comment (ayj) [reply galley/pkg/crd/validation/config.go]: `createOrUpdateWebhookConfig` already logs the per-call failure. Adding the retry interval here sounds good.

**Comment by ayj** [comment]
> time="19:44:25" level=debug msg="Image: \"istio-testing/proxyv2:istio-testing-distroless\" with ID \"sha256:1ebe92ea8625e30694ea4adb5c54e95d2b990a9934cf6a4e88af32365f3326ec\" not present on node \"istio-testing-control-plane\""

/test e2e-simpleTests-distroless_istio

**Comment by ayj** [comment]
> --- FAIL: TestCrdsRetryAsynchronously (0.13
>  store_test.go:491: Has 1 caches, Want 2 caches

/test unit-tests_istio

**Comment by ayj** [comment]
cherry-pick for release-1.3 => https://github.com/istio/istio/pull/17879
cherry-pick for release-1.2 => https://github.com/istio/istio/pull/17878

**Pull request merged**